chain/ethereum/ethereum.go

```diff
@@ -232,3 +234,71 @@ func BackfillFrom(ctx context.Context, client *ethclient.Client, addresses []com
 		return nil
 	}
 }
+
+//	HistoricalEvents queries past blocks for the events emitted by the given contract addresses.
+//	These events are provided in a channel and ready to be consumed by the caller.
+func HistoricalEvents(ctx context.Context, client *ethclient.Client, addresses []common.Address, fromBlock uint64, toBlock uint64) chan types.Log {
+	if fromBlock == toBlock {
+		return nil
+	}
+
+	if toBlock == 0 {
+		var err error
+		toBlock, err = client.BlockNumber(ctx)
+		if err != nil {
+			log.Error().Err(err).Msg("failed to get block number")
+		}
+	}
+
+	if fromBlock >= toBlock {
+		return nil
+	}
+
+	ch := make(chan types.Log, 1000)
+
+	go func(logs chan types.Log) {
+		defer close(logs)
+
+		//	Store failed queries for retry
+		failedQueries, err := ChunkedFilterLogs(ctx, client, addresses, fromBlock, toBlock, logs, nil)
+		if err != nil {
+			log.Error().Uint64("from", fromBlock).Uint64("to", toBlock).Msg("some intervals failed during backfill..")
+		}
+
+		for _, q := range failedQueries {
+			//	Retry failed queries one more time
+			fq, err := ChunkedFilterLogs(ctx, client, q.Addresses, q.FromBlock.Uint64(), q.ToBlock.Uint64(), logs, nil)
+			if err != nil {
+				log.Error().Uint64("from", q.FromBlock.Uint64()).Uint64("to", q.ToBlock.Uint64()).Msg("some intervals failed during backfill retry..")
+			}
+			for _, q2 := range fq {
+				log.Error().Uint64("from", q2.FromBlock.Uint64()).Uint64("to", q2.ToBlock.Uint64()).Msg("aborting failed backfill interval.")
+			}
+		}
+	}(ch)
+
+	return ch
+}
+
+// HistoricalEventsWithQueryParams queries past blocks for the events emitted by the given contract addresses.
+// These events are provided in a channel and ready to be consumed by the caller.
+// * fromBlock > 0 && numBlocks > 0 => Backfill from fromBlock to fromBlock+numBlocks
+// * fromBlock > 0 && numBlocks = 0 => Backfill from fromBlock to current latest block
+// * fromBlock = 0 && numBlocks > 0 => Backfill last numBlocks blocks
+func HistoricalEventsWithQueryParams(ctx context.Context, client *ethclient.Client, addresses []common.Address, fromBlock uint64, numBlocks uint64) chan types.Log {
+	switch {
+	case fromBlock > 0 && numBlocks > 0:
+		return HistoricalEvents(ctx, client, addresses, fromBlock, fromBlock+numBlocks)
+	case fromBlock > 0 && numBlocks == 0:
+		return HistoricalEvents(ctx, client, addresses, fromBlock, 0)
+	case fromBlock == 0 && numBlocks > 0:
+		toBlock, err := client.BlockNumber(ctx)
+		if err != nil {
+			log.Error().Err(err).Msg("failed to get block number")
+			return nil
+		}
+		return HistoricalEvents(ctx, client, addresses, toBlock-numBlocks, toBlock)
+	default:
+		return nil
```

Comment:
Return a nil channel can cause unexpexted result for caller who reads from this channel.
You can return a closed channel or return a nil channel with an error instead